File: a2a-samples/samples/go/agents/sign-and-verify-agent-card/test_client.go
package main

import (
	"context"
	"crypto/x509"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"log"
	"net/http"

	"github.com/a2aproject/a2a-go/v2/a2a"
	"github.com/a2aproject/a2a-go/v2/a2aclient"
	"github.com/a2aproject/a2a-go/v2/a2aclient/agentcard"
	"github.com/a2aproject/a2a-go/v2/a2asrv"
)

// keyProvider fetches and parses public key from JKU URL given key ID (keyID) and JKU URL.
func keyProvider(keyID, jku string) (any, error) {
	if keyID == "" {
		return nil, fmt.Errorf("expected keyID: string, but got empty string")
	}
	if jku == "" {
		return nil, fmt.Errorf("expected jku: string, but got empty string")
	}

	//nolint:gosec // JKU URL is dynamic by design per RFC 7515
	resp, err := http.Get(jku)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): %w", jku, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): status code %d", jku, resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): failed to read body: %w", jku, err)
	}

	var keys map[string]string
	if unmarshalErr := json.Unmarshal(body, &keys); unmarshalErr != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): invalid JSON response: %w", jku, unmarshalErr)
	}

	pemStr, ok := keys[keyID]
	if !ok || pemStr == "" {
		return nil, fmt.Errorf("invalid JWK key ID")
	}

	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM block for key ID %q", keyID)
	}

	pubKey, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse PKIX public key for key ID %q: %w", keyID, err)
	}

	return pubKey, nil
}

func displayAgentCard(card *a2a.AgentCard) {
	cardJSON, err := json.MarshalIndent(card, "", "  ")
	if err != nil {
		log.Fatalf("Failed to marshal agent card: %v", err)
	}
	log.Println(string(cardJSON))
}

func runTestClient() {
	ctx := context.Background()
	verifyCardSignature := createSignatureVerifier(keyProvider, []string{es256Alg})

	baseURL := serverURL

	log.Printf("Attempting to fetch public agent card from: %s%s", baseURL, a2asrv.WellKnownAgentCardPath)

	// Initialize A2ACardResolver using SDK helper
	resolver := agentcard.NewResolver(http.DefaultClient)

	// 1. Fetch public agent card without verifying signature
	_, err := resolver.Resolve(ctx, baseURL)
	if err != nil {
		log.Fatalf("Critical error fetching public agent card: %v", err)
	}
	log.Println("Successfully fetched public agent card without verification.")

	// 2. Fetch public agent card and verify signature
	publicCardVerified, err := resolver.Resolve(ctx, baseURL)
	if err != nil {
		log.Fatalf("Critical error fetching public agent card: %v", err)
	}
	if verifyErr := verifyCardSignature(publicCardVerified); verifyErr != nil {
		log.Fatalf("Failed to verify public agent card signature: %v", verifyErr)
	}
	log.Println("Successfully fetched public agent card with verification:")
	displayAgentCard(publicCardVerified)

	// Create A2A Client directly via verified public card
	client, err := a2aclient.NewFromCard(ctx, publicCardVerified)
	if err != nil {
		log.Fatalf("Failed to create A2A client: %v", err)
	}

	// 3. Fetch extended agent card without signature verification
	extendedCardUnverified, err := client.GetExtendedAgentCard(ctx, &a2a.GetExtendedAgentCardRequest{})
	if err != nil {
		log.Fatalf("Failed to get extended agent card: %v", err)
	}
	log.Println("Successfully fetched extended agent card without verification:")
	displayAgentCard(extendedCardUnverified)

	// 4. Fetch extended agent card and verify signature
	extendedCardVerified, err := client.GetExtendedAgentCard(ctx, &a2a.GetExtendedAgentCardRequest{})
	if err != nil {
		log.Fatalf("Failed to get extended agent card: %v", err)
	}
	if verifyErr := verifyCardSignature(extendedCardVerified); verifyErr != nil {
		log.Fatalf("Failed to verify extended agent card signature: %v", verifyErr)
	}
	log.Println("Successfully fetched extended agent card with verification:")
	displayAgentCard(extendedCardVerified)
}

File: a2a-samples/samples/python/agents/sign_and_verify_agent_card/test_client.py
import asyncio
import logging
import subprocess
import sys
import time

from pathlib import Path

import httpx
import pytest

from a2a.client import A2ACardResolver, ClientConfig, create_client
from a2a.helpers import display_agent_card
from a2a.types import GetExtendedAgentCardRequest
from a2a.utils.constants import (
    AGENT_CARD_WELL_KNOWN_PATH,
)
from a2a.utils.signing import create_signature_verifier
from cryptography.hazmat.primitives import serialization
from jwt.api_jwk import PyJWK


def _key_provider(key_id: str, jku: str) -> PyJWK | str | bytes:
    """Fetch and parse public key from JKU URL given key ID (key_id) and JKU URL."""
    if not isinstance(key_id, str) or not key_id:
        raise TypeError(f'Expected key_id: str, but got: {type(key_id).__name__} ({key_id!r})')
    if not isinstance(jku, str) or not jku:
        raise TypeError(f'Expected jku: str, but got: {type(jku).__name__} ({jku!r})')

    try:
        response = httpx.get(jku)
        response.raise_for_status()
    except httpx.HTTPError as err:
        raise ValueError(f'Failed to fetch public key from JKU URL ({jku}): {err}') from err

    keys = response.json()
    pem_data_str = keys.get(key_id)

    if not pem_data_str:
        raise ValueError('Invalid JWK Key ID.')

    return serialization.load_pem_public_key(pem_data_str.encode('utf-8'))


# Create a verifier function to validate AgentCard JWS signatures
verify_card_signature = create_signature_verifier(_key_provider, ['ES256'])


async def main() -> None:
    """Main function."""

    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)

    base_url = 'http://localhost:9999'

    async with httpx.AsyncClient() as httpx_client:
        # Initialize A2ACardResolver
        resolver = A2ACardResolver(
            httpx_client=httpx_client,
            base_url=base_url,
        )

        try:
            logger.info(
                'Attempting to fetch public agent card from: %s%s',
                base_url,
                AGENT_CARD_WELL_KNOWN_PATH,
            )
            # 1. Fetch public agent card without verifying signature
            _ = await resolver.get_agent_card()
            logger.info('Successfully fetched public agent card without verification.')

            # 2. Fetch public agent card and verify signature
            public_card_verified = await resolver.get_agent_card(
                signature_verifier=verify_card_signature,
            )
            logger.info('Successfully fetched public agent card with verification:')
            display_agent_card(public_card_verified)

        except Exception as e:
            logger.exception(
                'Critical error fetching public agent card.',
            )
            raise RuntimeError from e

        # Create Base Client directly via unified factory
        client = await create_client(
            agent=public_card_verified,
            client_config=ClientConfig(streaming=False),
        )

        # 3. Fetch extended agent card without signature verification
        extended_card_unverified = await client.get_extended_agent_card(
            GetExtendedAgentCardRequest()
        )
        logger.info('Successfully fetched extended agent card without verification:')
        display_agent_card(extended_card_unverified)

        # 4. Fetch extended agent card and verify signature
        extended_card_verified = await client.get_extended_agent_card(
            GetExtendedAgentCardRequest(),
            signature_verifier=verify_card_signature,
        )
        logger.info('Successfully fetched extended agent card with verification:')
        display_agent_card(extended_card_verified)

        await client.close()


@pytest.fixture(scope='session', autouse=True)
def start_server():
    server_path = Path(__file__).parent / '__main__.py'
    process = subprocess.Popen(  # noqa: S603
        [sys.executable, str(server_path)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        cwd=str(Path(__file__).parent),
    )
    # Wait a moment for the server to start
    time.sleep(1.5)

    if process.poll() is not None:
        raise RuntimeError(
            f'Server failed to start (exit code {process.poll()}). Port may already be in use.'
        )

    yield

    process.terminate()
    process.wait()


def test_client_workflow(text_query: str = 'Hi there!'):
    """
    This test function is intended to be used to test the client workflow
    for multiple queries
    """
    asyncio.run(main())


if __name__ == '__main__':
    asyncio.run(main())
